### PR TITLE
[AIC-py][eval] robust function calls

### PR DESCRIPTION
**Pull request by jonathanlastmileai**
[AIC-py][eval] robust function calls

main thing: call `run_thunk_safe()` for eval functions
- async, catch exceptions, timeout
- make a few important downstream changes, e.g. ID for logging/debugging

### CODE REVIEW

File: python/src/aiconfig/eval/common.py [hunk 1]
```diff
@@ -54,25 +54,25 @@ class EvaluationMetricMetadata(cu.Record, Generic[T_OutputDatum]):
...
     description: str
     best_value: MetricValue | None = None
     worst_value: MetricValue | None = None
     # e.g. {"substring": "hello", "case_sensitive": False}
     extra_metadata: dict[str, Any] = {}
+    id: str = "UNSET"
+
+    @model_validator(mode="before")
+    def set_id(cls, values: dict[str, Any]) -> dict[str, Any]:
+        values["id"] = cu.hash_id(f"{values['name']}{values['description']}{values['best_value']}{values['worst_value']}".encode("utf-8"))
```
Comment (rholinshead) [inline python/src/aiconfig/eval/common.py:67]: Won't this overwrite the `id` that the user has requested to be set?

File: python/src/aiconfig/eval/common.py [hunk 2]
```diff
@@ -54,25 +54,25 @@ class EvaluationMetricMetadata(cu.Record, Generic[T_OutputDatum]):
...
+        return values
+
+    def _serialize_extra_metadata(self) -> str:
+        return json.dumps(self.extra_metadata, sort_keys=True)
 
 
 class SampleMetricValue(cu.Record, Generic[T_OutputDatum]):
-    value: MetricValue
+    value: MetricValue | None
```
Comment (rholinshead) [inline python/src/aiconfig/eval/common.py:75]: What's the intention of a None value here?

File: python/src/aiconfig/eval/common.py (hunk 1 above)
Comment (jonathanlastmileai) [reply python/src/aiconfig/eval/common.py:67]: Yes and the intention is to set it automatically. I will figure out a different approach that prevents the user from setting it manually.

File: python/src/aiconfig/eval/common.py (hunk 2 above)
Comment (jonathanlastmileai) [reply python/src/aiconfig/eval/common.py:75]: Placeholder for errors

File: python/src/aiconfig/eval/lib.py [hunk 3]
```diff
@@ -86,23 +86,34 @@ def __str__(self) -> str:
...
+        def _ok_with_log(res_: Result[MetricValue, str]) -> MetricValue | None:
+            match res_:
+                case Ok(res):
+                    return res
+                case Err(e):
+                    LOGGER.error(f"Error evaluating {eval_params=}: {e}")
+                    return None
+
+        # TODO: figure out the right timeout
+        res_ = await cu.run_thunk_safe(_calculate(), timeout=1)
```
Comment (saqadri) [inline python/src/aiconfig/eval/lib.py:112]: is this 1 second? Would be good to add a quick comment, and ideally have timeout be passed in to the function

**Review by saqadri — APPROVED** [review]

**Pull request merged**